subprojects/android-test/ui-testing-core/src/main/kotlin/com/avito/android/test/matcher/TextMatchers.kt

```diff
@@ -0,0 +1,135 @@
+package com.avito.android.test.matcher
+
+import android.content.res.Resources
+import android.view.View
+import android.widget.TextView
+import androidx.test.espresso.matcher.BoundedDiagnosingMatcher
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+import org.hamcrest.StringDescription
+
+internal object TextMatchers {
+
+    internal fun withText(resourceId: Int) =
+        WithCharSequenceMatcher(
+            resourceId,
+            WithCharSequenceMatcher.TextViewMethod.GET_TEXT,
+        )
+
+    internal fun withText(text: String) =
+        withText(Matchers.`is`(text))
+
+    internal fun withText(stringMatcher: Matcher<String>) =
+        WithTextMatcher(stringMatcher)
+}
+
+internal class WithTextMatcher(
+    private val stringMatcher: Matcher<String>,
+) : BoundedDiagnosingMatcher<View, TextView>(TextView::class.java) {
+
+    override fun describeMoreTo(description: Description) {
+        description.appendText("view.getText() with or without transformation to match: ")
+        StringDescription().also {
+            stringMatcher.describeTo(it)
+            description.appendText(it.toString().highlightNbsp())
+        }
+    }
+
+    override fun matchesSafely(textView: TextView, mismatchDescription: Description): Boolean {
+        val text = textView.text.toString()
+        // The reason we try to match both original text and the transformed one is because some UI
+        // elements may inherit a default theme which behave differently for SDK below 19 and above.
+        // So it is implemented in a backwards compatible way.
+        if (stringMatcher.matches(text)) {
+            return true
+        }
+        mismatchDescription.appendText("view.getText() was ").appendValue(text.highlightNbsp())
+        if (textView.transformationMethod != null) {
+            val transformedText = textView.transformationMethod.getTransformation(text, textView)?.toString()
+            mismatchDescription.appendText(" transformed text was ")
+                .appendValue(transformedText?.highlightNbsp())
+            if (transformedText != null) {
+                return stringMatcher.matches(transformedText)
+            }
+        }
+        return false
+    }
+}
+
+internal class WithCharSequenceMatcher(
+    private val resourceId: Int,
+    private val method: TextViewMethod,
```

Comment:
inheritance will be more readable

Reply:
Unfortunately, these classes are final

Reply:
WithCharSequenceMatcher could be abstract and have two impls: HintWithCharSequenceMatcher, Text WithCharSequenceMatcher
Or I don't understand you